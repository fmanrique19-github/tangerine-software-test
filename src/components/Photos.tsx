import React, { useEffect } from 'react';
import { MDBCol, MDBListGroup, MDBListGroupItem, MDBJumbotron, MDBTable, MDBTableBody, MDBTableHead } from "mdbreact";
import runSaga from '../store/saga/injectSaga';
import { deletePhotos } from '../sagas/PhotosSaga';
import IPhotos, { IPhoto } from '../models/Photos';
import Spinner from './Spinner';

interface IFPhotos {
    photos: IPhotos
    md: number;
    className: string;
    title: string
    displayAlbums(): void;
};

const Photos = (props:  IFPhotos): JSX.Element => {
    
    const { photos, md, className, title } = props;

    useEffect(() => {

        return (() => {
            runSaga(deletePhotos);
        });
        
    }, []);

    const noPhotos = (): JSX.Element => {
        return (
            <MDBJumbotron>
                <h1 className="text-center">There is no photos available</h1>
            </MDBJumbotron>
        );
    }

    const showAlbums = () => {
        
        props.displayAlbums();
    }

    return (
        <MDBCol md={md as any} className={`tangerine-fixed ${className}`}>
            {
                !photos ? <Spinner /> : (photos.length === 0) ? noPhotos() :
                <MDBJumbotron className="photo-mobile">
                    <div onClick={ showAlbums } className="photo-mobile-close">
                        <svg id="i-close" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 32 32" width="32" height="32" fill="none" stroke="currentcolor" strokeLinecap="round" strokeLinejoin="round" strokeWidth="2">
                            <path d="M2 30 L30 2 M30 30 L2 2" />
                        </svg>
                    </div>
                    
                    <MDBListGroup>
                        <MDBListGroupItem href="#" className="text-center fixed-title" active key="0"><h3>Album's Photos - { title }</h3></MDBListGroupItem>
                        <div  className="scrollbar scrollbar-primary">
                        <MDBTable bordered>
                            <MDBTableHead color="gray-color">
                                <tr key="1">
                                <th>Title</th>
                                <th>Photo</th>
                                </tr>
                            </MDBTableHead>
                            <MDBTableBody>
                                {
                                    photos.map((photo: IPhoto) => {
                                        return(
                                            <tr key={ photo.id }>
                                            <td>{ photo.title }</td>
                                            <td><img src={ photo.thumbnailUrl } className="mx-auto d-block rounded" alt={ photo.title } /></td>
                                            </tr>
                                        );
                                    })
                                }
                            </MDBTableBody>
                        </MDBTable>
                        </div>
                    </MDBListGroup>
                </MDBJumbotron>
            }
        </MDBCol>
    );
}

export default Photos;
