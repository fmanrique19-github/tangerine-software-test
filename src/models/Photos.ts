export interface IPhoto {
    albumId: number,
    id: number,
    title: string,
    url: string,
    thumbnailUrl: string,
}

export default interface IPhotos extends Array<IPhoto>{};
