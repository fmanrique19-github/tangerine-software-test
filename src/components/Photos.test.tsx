import React from 'react';
import renderer from 'react-test-renderer';
import Photos from './Photos';
import mockStore from '../tests/helpers/mockStore';
import { Provider } from 'react-redux';
import { BrowserRouter as Router } from 'react-router-dom';

jest.mock('../store/saga/injectSaga');

describe('Photos Component', () => {

    it('Should render a spinner', () => {

        const props = {
            photos: null as any,
        };

        const tree = renderer
            .create(<Provider store={mockStore(props)}><Photos {...props} /></Provider>)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('Should render there is no photos available', () => {

        const props = {
            photos: [],
        };

        const tree = renderer
            .create(<Provider store={mockStore(props)}><Photos {...props} /></Provider>)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('Should render 3 photos', () => {

        const props = {
            photos: [
                {
                    "albumId": 1,
                    "id": 1,
                    "title": "accusamus beatae ad facilis cum similique qui sunt",
                    "url": "https://via.placeholder.com/600/92c952",
                    "thumbnailUrl": "https://via.placeholder.com/150/92c952"
                },
                {
                    "albumId": 1,
                    "id": 2,
                    "title": "reprehenderit est deserunt velit ipsam",
                    "url": "https://via.placeholder.com/600/771796",
                    "thumbnailUrl": "https://via.placeholder.com/150/771796"
                },
                {
                    "albumId": 1,
                    "id": 3,
                    "title": "officia porro iure quia iusto qui ipsa ut modi",
                    "url": "https://via.placeholder.com/600/24f355",
                    "thumbnailUrl": "https://via.placeholder.com/150/24f355"
                },
            ],
        };;

        const tree = renderer
            .create(<Provider store={mockStore(props)}><Router><Photos {...props} /></Router></Provider>)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

});
