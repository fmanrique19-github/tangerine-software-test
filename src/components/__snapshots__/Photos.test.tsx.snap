// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Photos Component Should render 3 photos 1`] = `
<div
  className="col tangerine-fixed undefined"
  data-test="col"
>
  <div
    className="jumbotron photo-mobile"
    data-test="jumbotron"
  >
    <div
      className="photo-mobile-close"
      onClick={[Function]}
    >
      <svg
        fill="none"
        height="32"
        id="i-close"
        stroke="currentcolor"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth="2"
        viewBox="0 0 32 32"
        width="32"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M2 30 L30 2 M30 30 L2 2"
        />
      </svg>
    </div>
    <ul
      className="list-group"
      data-test="list-group"
    >
      <a
        className="list-group-item text-center fixed-title active list-group-item-undefined"
        data-test="list-group-item"
        href="#"
      >
        <h3>
          Album's Photos - 
        </h3>
      </a>
      <div
        className="scrollbar scrollbar-primary"
      >
        <div
          className=""
          data-test="table"
          style={
            Object {
              "maxHeight": null,
            }
          }
        >
          <table
            className="table table-bordered"
          >
            <thead
              className="gray-color"
              data-test="table-head"
            >
              <tr>
                <th>
                  Title
                </th>
                <th>
                  Photo
                </th>
              </tr>
            </thead>
            <tbody
              data-test="table-body"
            >
              <tr>
                <td>
                  accusamus beatae ad facilis cum similique qui sunt
                </td>
                <td>
                  <img
                    alt="accusamus beatae ad facilis cum similique qui sunt"
                    className="mx-auto d-block rounded"
                    src="https://via.placeholder.com/150/92c952"
                  />
                </td>
              </tr>
              <tr>
                <td>
                  reprehenderit est deserunt velit ipsam
                </td>
                <td>
                  <img
                    alt="reprehenderit est deserunt velit ipsam"
                    className="mx-auto d-block rounded"
                    src="https://via.placeholder.com/150/771796"
                  />
                </td>
              </tr>
              <tr>
                <td>
                  officia porro iure quia iusto qui ipsa ut modi
                </td>
                <td>
                  <img
                    alt="officia porro iure quia iusto qui ipsa ut modi"
                    className="mx-auto d-block rounded"
                    src="https://via.placeholder.com/150/24f355"
                  />
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </ul>
  </div>
</div>
`;

exports[`Photos Component Should render a spinner 1`] = `
<div
  className="col tangerine-fixed undefined"
  data-test="col"
>
  <div
    className="d-flex justify-content-center"
  >
    <div
      className="spinner-border"
      role="status"
    >
      <span
        className="sr-only"
      >
        Loading...
      </span>
    </div>
  </div>
</div>
`;

exports[`Photos Component Should render there is no photos available 1`] = `
<div
  className="col tangerine-fixed undefined"
  data-test="col"
>
  <div
    className="jumbotron"
    data-test="jumbotron"
  >
    <h1
      className="text-center"
    >
      There is no photos available
    </h1>
  </div>
</div>
`;
